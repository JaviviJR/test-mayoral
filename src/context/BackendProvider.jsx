import { useEffect, useState } from "react";
import { BackendContext } from "./BackendContext";

const URI_PRODUCTS = 'http://localhost:9005/products';

function BackendProvider({ children }) {
    const [products, setProducts] = useState([]);

    useEffect(() => {
        getProducts();
    }, []);

    const getProducts = async (queryString) => {
        const uri = queryString ? `${URI_PRODUCTS}?${queryString}` : URI_PRODUCTS;
        const response = await fetch(uri);
        const data = await response.json();
        setProducts(data);
    }

    const orderProducts = (order) => {
        const productSorted = [...products];

        if (order === 'ASC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) > (b.offer ? b.offer : b.price));
        } else if (order === 'DESC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) < (b.offer ? b.offer : b.price));
        }

        setProducts(productSorted);
        
    }

    const searchProduct = (search) => {
        const queryString = `name_like=${search}`;
        getProducts(queryString);
    }

    return (
        <BackendContext.Provider value={{
            products,
            orderProducts,
            searchProduct
        }}>
            { children }
        </BackendContext.Provider>
    );
}

export default BackendProvider;