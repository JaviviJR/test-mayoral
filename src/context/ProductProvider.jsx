import { useState } from "react";
import { ProductContext } from "./ProductContext";

const URI_PRODUCTS = 'http://localhost:9005/products';

function ProductProvider({ children }) {
    const [products, setProducts] = useState([]);

    const getProducts = async (search) => {
        const uri = search ? `${URI_PRODUCTS}?name_like=${search}` : URI_PRODUCTS;
        const response = await fetch(uri);
        const data = await response.json();
        setProducts(data);
    }

    const orderProducts = (order) => {
        const productSorted = [...products];

        if (order === 'ASC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) > (b.offer ? b.offer : b.price));
        } else if (order === 'DESC') {
            productSorted.sort((a, b) => (a.offer ? a.offer : a.price) < (b.offer ? b.offer : b.price));
        }

        setProducts(productSorted);
        
    }

    return (
        <ProductContext.Provider value={{
            products,
            orderProducts,
            getProducts
        }}>
            { children }
        </ProductContext.Provider>
    );
}

export default ProductProvider;