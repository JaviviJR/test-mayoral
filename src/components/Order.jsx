import { useContext } from "react";
import { ProductContext } from "../context/ProductContext";

function Order() {
    const { orderProducts } = useContext(ProductContext);

    return (
        <>
            <div 
                className="h1"
                onClick={ () => orderProducts('DESC') }
                style={{ cursor:"pointer" }}
            >
                -
            </div>
            <div 
                className="h1"
                onClick={ () => orderProducts('ASC') }
                style={{ cursor:"pointer" }}
            >
                +
            </div>
        </>
    );
}

export default Order;