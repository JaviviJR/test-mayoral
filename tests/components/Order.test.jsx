import { render, screen, fireEvent, waitFor } from "@testing-library/react"
import { ProductContext } from "../../src/context/ProductContext";
import Order from "../../src/components/Order";

describe('test <InputSearch />', () => {
    const orderProducts = jest.fn();
    
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe coincidir con el snapshot', () => {
        const { container } = render(
            <ProductContext.Provider value={{
                orderProducts
            }}>
                <Order />
            </ProductContext.Provider>
        );
        expect(container).toMatchSnapshot();
    })

    test('debe llamar orderProducts("ASC") al hacer click en "+"', () => {
        render(
            <ProductContext.Provider value={{
                orderProducts
            }}>
                <Order />
            </ProductContext.Provider>
        );

        const input = screen.getByText('+');
        fireEvent.click(input);
        expect(orderProducts).toHaveBeenCalledTimes(1);
        expect(orderProducts).toHaveBeenCalledWith('ASC');
    })

    test('debe llamar orderProducts("DESC") al hacer click en "-"', () => {
        render(
            <ProductContext.Provider value={{
                orderProducts
            }}>
                <Order />
            </ProductContext.Provider>
        );

        const input = screen.getByText('-');
        fireEvent.click(input);
        expect(orderProducts).toHaveBeenCalledTimes(1);
        expect(orderProducts).toHaveBeenCalledWith('DESC');
    })
})