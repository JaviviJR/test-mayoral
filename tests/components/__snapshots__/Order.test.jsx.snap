// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <InputSearch /> debe coincidir con el snapshot 1`] = `
<div>
  <div
    class="h1"
    style="cursor: pointer;"
  >
    -
  </div>
  <div
    class="h1"
    style="cursor: pointer;"
  >
    +
  </div>
</div>
`;
